library/sha256.c

```diff
@@ -143,6 +243,132 @@ static const uint32_t K[] =
     0x90BEFFFA, 0xA4506CEB, 0xBEF9A3F7, 0xC67178F2,
 };
 
+#endif
+
+#if defined(MBEDTLS_SHA256_USE_A64_CRYPTO_IF_PRESENT) || \
+    defined(MBEDTLS_SHA256_USE_A64_CRYPTO_ONLY)
+
+#if defined(MBEDTLS_SHA256_USE_A64_CRYPTO_ONLY)
+#  define mbedtls_internal_sha256_process_many_a64_crypto mbedtls_internal_sha256_process_many
+#  define mbedtls_internal_sha256_process_a64_crypto      mbedtls_internal_sha256_process
+#endif
+
+static size_t mbedtls_internal_sha256_process_many_a64_crypto(
+                  mbedtls_sha256_context *ctx, const uint8_t *msg, size_t len )
+{
+    uint32x4_t abcd = vld1q_u32( &ctx->state[0] );
+    uint32x4_t efgh = vld1q_u32( &ctx->state[4] );
+
+    size_t processed = 0;
+
+    for( ;
+         len >= SHA256_BLOCK_SIZE;
+         processed += SHA256_BLOCK_SIZE,
+               msg += SHA256_BLOCK_SIZE,
+               len -= SHA256_BLOCK_SIZE )
+    {
+        uint32x4_t tmp, abcd_prev;
+
+        uint32x4_t abcd_orig = abcd;
+        uint32x4_t efgh_orig = efgh;
+
+        uint32x4_t sched0 = vld1q_u32( (const uint32_t *)( msg + 16 * 0 ) );
+        uint32x4_t sched1 = vld1q_u32( (const uint32_t *)( msg + 16 * 1 ) );
+        uint32x4_t sched2 = vld1q_u32( (const uint32_t *)( msg + 16 * 2 ) );
+        uint32x4_t sched3 = vld1q_u32( (const uint32_t *)( msg + 16 * 3 ) );
+
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__  /* Will be true if not defined */
+                                               /* Untested on BE */
+        sched0 = vreinterpretq_u32_u8( vrev32q_u8( vreinterpretq_u8_u32( sched0 ) ) );
+        sched1 = vreinterpretq_u32_u8( vrev32q_u8( vreinterpretq_u8_u32( sched1 ) ) );
+        sched2 = vreinterpretq_u32_u8( vrev32q_u8( vreinterpretq_u8_u32( sched2 ) ) );
+        sched3 = vreinterpretq_u32_u8( vrev32q_u8( vreinterpretq_u8_u32( sched3 ) ) );
+#endif
+
+        /* Rounds 0 to 3 */
+        tmp = vaddq_u32( sched0, vld1q_u32( &K[0] ) );
```

Comment:
We can see, for every loop, we should load the K[0], K[1]...K[15]. can we try to load once and use it every loop? maybe it will be limited by the number of registers. but not sure whether we have exhaused all of the registers now.

Reply:
That is an optimisation question I think the compiler should be able to sort out, for whichever uarch it is compiling for

Reply:
Compiler can allocate registers automatically, but seems here we will call vld1q_u32() every loop, does that mean every loop we have to load the Ks from memory to registers? If we can keep the registers, then we can save several "Load" instructions every loop.

Reply:
However, I will see if reordering some of these instructions improves performance

Reply:
And pre-loading with

```
    uint32x4_t k0 = vld1q_u32( &K[0] );
    uint32x4_t k4 = vld1q_u32( &K[4] );
    uint32x4_t k8 = vld1q_u32( &K[8] );
    uint32x4_t k12 = vld1q_u32( &K[12] );
```

made < 0.1% difference, sometimes faster, sometimes slower - i.e. no impact

Reply:
> seems here we will call `vld1q_u32()` every loop

Sorry, I missed this - `vld1q_u32()` is an intrinsic function that just means "load four 32-bit values" - it directly translates into `ldr	qN, [...]`

Reply:
How about pre-loading the Ks before the loop ```
"    for( ;
         len >= SHA256_BLOCK_SIZE;
         processed += SHA256_BLOCK_SIZE,
               msg += SHA256_BLOCK_SIZE,
               len -= SHA256_BLOCK_SIZE )"
```,  Is there any improvement?

Reply:
@xkqian Yes, with the `k0` to `k12` preloads I put those before that loop. It made < 0.1% difference. I didn't try with all the `k` values, but I don't think it would have helped

Reply:
How about the big data size? 
From the assembly code you provided, with current code, every loop it will have 4 extra instructions, such as : "ldr	q6, [x27, lCPI5_0@PAGEOFF], ldr	q6, [x28, lCPI5_1@PAGEOFF], ldr	q6, [x23, lCPI5_2@PAGEOFF], ldr	q6, [x24, lCPI5_3@PAGEOFF]".
If we increase the data size, which means more loops(n), then if we pre-load it, can we save 4*n instrcutions idealy?

Reply:
Yes, but that's what I did. I put the four `k0` to `k12` as preloads before that loop. It made < 0.1% difference to performance

Reply:
> Yes, but that's what I did. I put the four `k0` to `k12` as preloads before that loop. It made < 0.1% difference to performance

From the assembly code: 
```
"	ldr	q6, [x27, lCPI5_0@PAGEOFF]
	add.4s	v6, v2, v6
	mov.16b	v7, v0
	sha256h.4s	q7, q1, v6
	mov.16b	v16, v1
	sha256h2.4s	q16, q0, v6"
```
This is one strange result cause every 6 instructions has 1 LDR instrcution. and LDR has big latency(even reach to 4(2) on N1).  Anyway, if  the pre-loading can't benefit us, I think it's ok to leave the code as it is.